syntax = "proto2";

package pmidgserialized;

message GameState
{
    repeated Entity entities = 1;
    repeated System active_systems = 2;
    required Instance instance = 3;
}

message Entity
{
    required int32 id = 1;
    repeated Component components = 2;
}

message System
{
    required string name = 1;
}

message Instance
{
    required string tilemap = 1;
}

message Component
{
    required string name = 1;
    repeated ComponentValueString string_values = 2;
    repeated ComponentValueBool bool_values = 3;
    repeated ComponentValueFloat float_values = 4;
}

message ComponentValueString
{
    required string value_name = 1;
    required string value = 2;
}

message ComponentValueBool
{
    required string value_name = 1;
    required bool value = 2;
}

message ComponentValueFloat
{
    required string value_name = 1;
    required float value = 2;
}